packages/lsp/src/adapters/editorAdapter.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { CodeEditor } from '@jupyterlab/codeeditor';
+
+import {
+  IConfigurableExtension,
+  IEditorExtensionFactory
+} from '@jupyterlab/codemirror';
+
+import { IDisposable } from '@lumino/disposable';
+
+import { Signal } from '@lumino/signaling';
+import { WidgetLSPAdapter } from './adapter';
+
+/**
+ * The CodeEditor.IEditor adapter.
+ */
+export class EditorAdapter implements IDisposable {
+  /**
+   * Instantiate a new EditorAdapter.
+   *
+   * @param options The instantiation options for a EditorAdapter.
+   */
+  constructor(options: EditorAdapter.IOptions) {
+    this._widgetAdapter = options.widgetAdapter;
+    this._extensions = options.extensions;
+
+    void options.editorReady().then(editor => {
+      this._injectExtensions(editor);
+    });
+  }
+
+  /**
+   * Whether the handler is disposed.
+   */
+  isDisposed: boolean;
+
+  /**
+   * Dispose the handler.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this.isDisposed = true;
+    Signal.clearData(this);
+  }
+
+  /**
+   * Setup the editor.
+   */
+  private _injectExtensions(editor: CodeEditor.IEditor): void {
+    if (editor.isDisposed) {
+      return;
+    }
+
+    this._extensions.forEach(factory => {
+      const ext = factory.factory({
+        path: this._widgetAdapter.widget.context.path,
+        editor: editor,
+        widgetAdapter: this._widgetAdapter,
+        model: editor.model,
+        inline: true
+      });
+
+      if (!ext) {
+        return;
+      }
+
+      editor.injectExtension(ext.instance(editor));
+    });
+  }
+
+  private _widgetAdapter: WidgetLSPAdapter<any>;
+  private _extensions: EditorAdapter.ILSPEditorExtensionFactory[];
+}
+
+/**
+ * A namespace for EditorAdapter `statics`.
+ */
+export namespace EditorAdapter {
+  /**
+   * Instantiation options for `EditorAdapter`.
+   */
+  export interface IOptions {
+    /**
+     * Promise resolving when the editor is ready.
+     */
+    editorReady(): Promise<CodeEditor.IEditor>;
+
+    /**
+     * The widget lsp adapter.
+     */
+    widgetAdapter: WidgetLSPAdapter;
+
+    /**
+     * The list of CodeMirror extension factories
+     */
+    extensions: ILSPEditorExtensionFactory[];
+  }
+
+  export interface ILSPEditorExtensionFactory
+    extends Omit<IEditorExtensionFactory<any>, 'factory'> {
+    /**
+     * Extension factory.
+     *
+     * @param options
+     * @returns The extension builder or null if the extension is not active for that document
+     */
+    readonly factory: (
+      options: IFactoryOptions
+    ) => IConfigurableExtension<any> | null;
+  }
+
+  export interface IFactoryOptions extends IEditorExtensionFactory.IOptions {
+    /**
+     * A path to a source file.
+     */
+    path: string;
+    /**
+     * The code editor.
+     */
+    editor: CodeEditor.IEditor;
```

Comment:
The last problem with this is that we have something called `Document.IEditor` which is what the adapter and virtual document APIs operates on (for example `transformFromEditorToRoot` and `getIndexAt` methods). `Document.IEditor` replaced `CodeEditor.IEditor` on the LSP APIs when virtual notebook feature was introduced in https://github.com/jupyterlab/jupyterlab/pull/12554.

The lack of mapping from `CodeEditor.IEditor` to `Document.IEditor` prevents the direct use of `CodeEditor.IEditor` on adapter and virtual document methods. To workaround it  I have to iterate `adapter.editors` to find which `Document.IEditor` wraps the `CodeEditor.IEditor` at hand, which is potentially inefficient for long notebooks. Btw, I use "editorAccessor" to refer to `Document.IEditor` as this gets very confusing otherwise.

Conceptually the solution is simple we can: pass `editor: Document.IEditor` instead of `editor: CodeEditor.IEditor` but this may be non-trivial to implement. Alternatively, if there is an oracle with a method `getEditorAccessorForEditor(e: CodeEditor.IEditor): Document.IEditor` we can keep it as-is.

CC @fcollonval for thoughts

Reply:
> Conceptually the solution is simple we can: pass editor: Document.IEditor instead of editor: CodeEditor.IEditor but this may be non-trivial to implement.

I think it is fine, we can pass the `Document.IEditor` instead of the `CodeEditor.IEditor`. See 420b76db8824b64ebe6e73cdcb17560a0e6e1f85